### ISSUE TITLE
ABCI: add CheckBlock into Application to allow app checking block before prevote

### ISSUE THREAD

**Issue by cjqpker** [tendermint/tendermint#3759]
In below senarios we may need this feature:

1、We don't allow invalid txs included in each block so we need to run block before prevote , the same way as btc and ethereum

2、Our app need to support sharding modules and each sharding module has it's own stateroot respectively so we need to check sharding state root hash in application

**Comment by melekes** [comment]
See discussions in #2384 and #2175

**Comment by cjqpker** [comment]
> See discussions in #2384 and #2175

No aboud invalid txs, we want to verify the whole block including block header.

**Comment by ebuchman** [comment]
We already do validate the block before prevoting, including all fields in the block header. The only thing we don't do is actually execute it.

**Comment by cjqpker** [comment]
> We already do validate the block before prevoting, including all fields in the block header. The only thing we don't do is actually execute it.

Block is validated in consensus process, not in abci application. What if I want to check some header fields in abci application when validation needs data from abci app?

**Comment by ebuchman** [comment]
Interesting. So something like a `CheckBeginBlock` basically? Or you'd want the entire block? Would probably have to be some kind of optional ABCI feature, since it's not clear everyone needs/wants this yet. Bringing optionality to the ABCI is something we've been considering but needs to be done carefully so as to not over complicate

**Comment by cjqpker** [comment]
> Interesting. So something like a `CheckBeginBlock` basically? Or you'd want the entire block? Would probably have to be some kind of optional ABCI feature, since it's not clear everyone needs/wants this yet. Bringing optionality to the ABCI is something we've been considering but needs to be done carefully so as to not over complicate

We tried to expose the block txs array in BeginBlock and verify signatures parallelly and cache the verification results, so DeliverTx can directlly use the cache and the tps is doubled.

Maybe ABCI should be as simple as possible, but lacking optionality limits its expansibility.

**Comment by melekes** [comment]
> We tried to expose the block txs array in BeginBlock and verify signatures parallelly and cache the verification results, so DeliverTx can directlly use the cache and the tps is doubled.

Right. We have an issue for that - https://github.com/tendermint/tendermint/issues/2901

**Comment by melekes** [comment]
related: https://forum.cosmos.network/t/introduction-to-abcix-an-extension-of-abci-with-greater-flexibility-and-security/3771
